src/main/java/org/icgc/argo/program_service/services/ego/EgoService.java

```diff
@@ -0,0 +1,253 @@
+@Slf4j
+@Service
+public class EgoService {
+  private final EgoRESTClient egoClient;
+  private final ProgramEgoGroupRepository programEgoGroupRepository;
+  private final ProgramConverter programConverter;
+  private RSAPublicKey egoPublicKey;
+
+  @Autowired
+  public EgoService(
+    @NonNull ProgramEgoGroupRepository programEgoGroupRepository,
+	  @NonNull ProgramConverter programConverter,
+    @NonNull EgoRESTClient restClient
+      ) {
+    this.programEgoGroupRepository = programEgoGroupRepository;
+    this.programConverter = programConverter;
+    this.egoClient = restClient;
+  }
+
+  @Autowired
+  private void setEgoPublicKey() {
+    this.egoPublicKey = egoClient.getPublicKey();
+  }
+
+
+  public Optional<EgoToken> verifyToken(String jwtToken) {
+    try {
+      Algorithm algorithm = Algorithm.RSA256(this.egoPublicKey, null);
+      JWTVerifier verifier = JWT.require(algorithm)
+        .withIssuer("ego")
+        .build(); //Reusable verifier instance
+      val jwt = verifier.verify(jwtToken);
+      return parseToken(jwt);
+    } catch (JWTVerificationException | NullPointerException e) {
+      // Handle NPE defensively for null JWT.
+      return Optional.empty();
+    }
+  }
+
+  public Optional<EgoToken> parseToken(DecodedJWT jwt) {
+    try {
+      EgoToken egoToken = new EgoToken(jwt, jwt.getClaim("context").as(Context.class));
+      return Optional.of(egoToken);
+    } catch (JWTDecodeException exception){
+      //Invalid token
+      return Optional.empty();
+    }
+  }
+
+
+  //TODO: add transactional. If there are more programdb logic in the future and something fails, it will be able to roll back those changes.
+  public void setUpProgram(@NonNull ProgramEntity program, @NonNull Collection<String> adminEmails) {
+    val programPolicy = egoClient.createEgoPolicy("PROGRAM-" + program.getShortName());
+    val dataPolicy = egoClient.createEgoPolicy("PROGRAMDATA-" + program.getShortName());
+
+    Stream.of(UserRole.values())
+      .filter(role -> !role.equals(UserRole.UNRECOGNIZED))
+      .forEach(
+        role -> {
+          val group = ensureGroupExists(program, role);
+          egoClient.assignPermission(group, programPolicy, getProgramMask(role));
+          egoClient.assignPermission(group, dataPolicy, getDataMask(role));
+          saveGroupIdForProgramAndRole(program, role, group.getId());
+
+        }
+      );
+
+    adminEmails.forEach(email -> initAdmin(email, program));
+  }
+
+  public String getProgramMask(UserRole role) {
+    switch(role) {
+    case ADMIN: return "WRITE"; // return ADMIN
+    case CURATOR: return "WRITE"; // check this with spec
+    case SUBMITTER: return "READ";
+    case COLLABORATOR: return "READ";
+    case BANNED: return "DENY";
+    default: throw new IllegalArgumentException("Invalid role " + role.toString());
+    }
+  }
+
+  public String getDataMask(UserRole role) {
+    switch(role) {
+    case ADMIN: // return "ADMIN";
```

Comment:
Whats with the empty and comment?

Reply:
Ego didn't have an ADMIN role last time I checked.
When it does, that's what we need to return there. :-)